### PR TITLE
Use the official Elasticsearch Go client.

### PR DESCRIPTION
**Pull request by nkvoll**
This replaces the implementations of the `elasticsearch/client.Client` interfaces for v6 and v7 with a single implementation that spans v6, v7 and v8 APIs using the official Go client (https://github.com/elastic/go-elasticsearch).

Since the official client is still a low-level client, we still keep the `client/model.go` to deal with marshalling/unmarshalling.

### CODE REVIEW

**Comment by sebgl** [comment]
I'll get a deeper look at this next week. In the background of my head I'm trying to judge the trade-offs between the additional complexity/verbosity this PR brings vs. its benefits.

A pros/cons listing draft:

Cons:
* all API responses are generic, we do not benefit from any structured model per version of Elasticsearch that would help us detect breaking changes at compile time
* same for requests payload
* some API calls are missing (but that's short-term a bug that will be fixed)
* there is no default retry logic for all ES versions
* we need to switch over the different clients for every single API call
* we need to maintain an up-to-date list of vendored clients (for each version)

Pros:
* eat your own food: Elastic provides an official go client, why building our own?
* opportunity to improve the official client along the way (response model & retry logic are the most important missing bits IMO)
* API that may not exist in a given ES version are caught at compile-time
* we're forced to think about the various ES versions in the code

I tend to think the pros outweigh the cons in the long-term, especially if we can improve the client in parallel.
Anything I missed @nkvoll?

File: pkg/controller/elasticsearch/client/official_log.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+// esLogger implements estransport.Logger using the controller-runtime logging infrastructure
+type esLogger struct{}
+
+var (
+	_ estransport.Logger = &esLogger{}
+)
+
+// LogRoundTrip should not modify the request or response, except for consuming and closing the body.
+// Implementations have to check for nil values in request and response.
+func (l *esLogger) LogRoundTrip(
```
Comment (karmi) [inline pkg/controller/elasticsearch/client/official_log.go]: Maybe you can (re)use the [`estransport.JSONLogger`](https://github.com/elastic/go-elasticsearch/blob/master/estransport/logger.go) here?

File: pkg/controller/elasticsearch/driver/driver.go [hunk 2]
```diff
@@ -155,14 +155,19 @@ func (d *defaultDriver) Reconcile() *reconciler.Results {
 
 	warnUnsupportedDistro(resourcesState.AllPods, d.ReconcileState.Recorder)
 
+	observerESClient, err := d.newElasticsearchClient(
+		resourcesState,
+		internalUsers.ControllerUser,
+		*min,
```
Comment (david-kow) [inline pkg/controller/elasticsearch/driver/driver.go]: What happens when we use v6 client to talk to v7 nodes? This will happen during upgrades, since, IIUC, we pick min version of all pods.

**Comment by karmi** [comment]
> there is no default retry logic for all ES versions

While the [support for retries](https://github.com/elastic/go-elasticsearch/commit/7f37b7b9a63588ce18b365d4e07917dbda8e55a0) is only in the `master` branch at the moment, it will be merged into the `7.x` and `6.x` branches soon. I wanted to "bake it" for a while in `master`, before merging it to version branches, and making it part of a release.

Note, the same applies to the new [support for connection pooling](https://github.com/elastic/go-elasticsearch/pull/95), and the yet-to-be-added support for node discovery ("sniffing").

A note about the "response model", which we have talked about in chats. While the specification for request and response _bodies_ — note: we _do have_ a [specification](https://github.com/elastic/elasticsearch/tree/master/rest-api-spec/src/main/resources/rest-api-spec/api) for the API "surface" — is something we want (and perhaps need) to have at some point, it is nowhere near being on the radar. Also, it's a highly complex problem to solve, which is one reason why we don't have it yet. The structure of the request/body is highly variable, and there's no clear way of encoding it, as we did with the API "surface". Note, that all clients in strongly-typed languages (Java, .NET, Go) maintain this layer manually, which is terribly time-consuming, laborious and error-prone. It _must_ be automated in some way, if we want to provide strong guarantees.

File: pkg/controller/elasticsearch/driver/driver.go (hunk 2 above)
Comment (karmi) [reply pkg/controller/elasticsearch/driver/driver.go]: > What happens when we use v6 client to talk to v7 nodes?

For most API calls, it will go through fine, but technically speaking, it is an unsupported scenario. Elasticsearch might add support for a request-specific version of the API to use (eg. in the header, as part of content negotiation), so that's something which might help.

File: pkg/controller/elasticsearch/client/official_log.go (hunk 1 above)
Comment (nkvoll) [reply pkg/controller/elasticsearch/client/official_log.go]: The underlying logger does the serializing of the arguments JSON in our case, so I don't think the JSONLogger would be applicable (e.g resulting in double-serialized since we can't pass it already-seriailized data).

Comment (nkvoll) [reply pkg/controller/elasticsearch/client/official_log.go]: The underlying logger we use would serialize the arguments to JSON, resulting in double-encoding instead of embedding some partial JSON. I suppose we could work around it by wrapping the JSONLogger result in a struct we provide a custom marshaller for, but it seems easier to do it this way for now

**Comment by nkvoll** [comment]
Longer term re our client usage I think we should look into putting it behind an internal API that makes more high level statements and questions, such as: "prepare these nodes for restart/shutdown/removal", "is it safe to restart / shutdown / remove these nodes now"?, which would potentially be implemented on a per-version basis, avoiding this switching. Maybe we could even work with the ES team to see if these could translate more or less directly into APIs to call on that side as well. The idea is not fully baked yet so treat it for what it is -- just an idea at this point.

As for compatibility and which version of the ES client to use... It's not a simple question to answer. We're talking to ES via a Service, so we don't know which version we will be talking with for any particular request by default. We could connect to specific pods directly to avoid this ambiguity, but that is far outside the scope of this PR, but worth seeing where it will lead us.

File: pkg/controller/elasticsearch/client/official_log.go (hunk 1 above)
Comment (karmi) [reply pkg/controller/elasticsearch/client/official_log.go]: Argh, right, I've missed that you pass a structure to the underlying logger. In this case a custom Logger implementation is definitely the right solution.

**Comment by karmi** [comment]
I forgot to mention the [**`esutil.JSONReader`**](https://github.com/elastic/go-elasticsearch/blob/master/_examples/encoding/jsonreader.go) helper in the Go client, which at least allows for easy encoding of common types like `map[string]interface{}`.

**Comment by pebrc** [comment]
Potential future work related to this https://github.com/elastic/cloud-on-k8s/issues/1016

File: pkg/controller/elasticsearch/client/official_api.go [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+		},
+		ES8: func(es *elasticsearch8.Client) (*esapi8.Response, error) {
+			return es.Cluster.Health(
+				es.Cluster.Health.WithContext(ctx),
+			)
+		},
+		ResponseHandler: esErrorOrDecodeJSON(&result),
+	}); err != nil {
+		return result, err
+	}
```
Comment (sebgl) [inline pkg/controller/elasticsearch/client/official_api.go]: So I guess an alternative to the `switch` in every call would be to have a `client` interface implemented by `clientv6`, `clientv7`, `clientv8`? Since they use different types, every single one would need to implement all functions of the generic client interface, and maybe panic or return an error if irrelevant (eg. zen1 on clientv7).
Which is pretty similar to the approach we had before this PR I think, but will be more verbose since need to implement all functions (cannot have clientv7 embed clientv6).

The major benefit I see is that we cannot accidentally forget a `case` in all switches: all clients will have to implement the interface. Also it avoids the `if versionX != nil` statements.

I don't think it should prevent this from being merged. Could be nice as a follow-up issue?

File: pkg/controller/elasticsearch/client/official.go [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+// ESError returns the Elasticsearch API response as an error
+func (r *Response) ESError() error {
+	r.esErrorOnce.Do(func() {
+		r.esError = newOfficialAPIError(r)
+	})
+	return r.esError
+}
+
+// officialAPIError is a non 2xx response from the Elasticsearch API
+type officialAPIError struct {
```
Comment (sebgl) [inline pkg/controller/elasticsearch/client/official.go]: `official` being used in multiple places (also in file names) feels a bit weird to me (naming 🤷‍♂).
Why not just `APIError`? If we need to "wrap" things in our own structs I guess we could move everything related to the go-elasticsearch client in its own sub-package.

File: pkg/controller/elasticsearch/driver/driver.go (hunk 2 above)
Comment (sebgl) [reply pkg/controller/elasticsearch/driver/driver.go]: We already have a logic of switching from the generic cluster service URL to a particular Pod URL: https://github.com/elastic/cloud-on-k8s/blob/38ff3d3f4b4ff63f2db10afb6cd816fb4ebc8056/pkg/controller/elasticsearch/services/services.go#L107-L115
We needed it to handle the http/https migration. I think we could use something similar to handle version upgrades. If there are pods in 2 different versions, build a client for the lower version and use it to talk to that pods directly. So far that's not really a big problem though (when it is, we'll retry at next reconciliation anyway).

File: pkg/controller/elasticsearch/client/client_test.go [hunk 5]
```diff
@@ -311,6 +273,8 @@ func TestClientGetNodes(t *testing.T) {
 			Request:    req,
 		}
 	})
```
Comment (sebgl) [inline pkg/controller/elasticsearch/client/client_test.go]: These tests seem quite irrelevant now. We're mostly testing the mock here.
I think we can trust the go-elasticsearch client and not double-test the right function calls the right url?

However, I would maybe expect some tests for `official.go` and `official_api.go`.

**Comment by pebrc** [comment]
@nkvoll I am closing this as it has been open for 6 months now. Please reopen when you want to pick this up again.
